### PR TITLE
[STRIPES-859] Re-export type declarations from stripes-types

### PR DESCRIPTION
**Pull request by ncovercash**
# [Jira STRIPES-859](https://issues.folio.org/browse/STRIPES-859)

This PR adds `index.d.ts` files to re-export types stored in `stripes-types`.  Initially, this consists of typings for `stripes-components`, `stripes-core`, `stripes-final-form`, and `stripes-smart-components` — all other modules will have no associated type information, as it was previously.

For convenience, here is a link to [stripes-types](https://github.com/folio-org/stripes-types/tree/v1.0.1).

### What is in `stripes-types` so far?

For modules `stripes-components`, `stripes-core`, and `stripes-smart-components`, all declarations are denoted as `any`, disabling type checking and having no effect, e.g.:

```ts
export const RootContext: any;
export const withRoot: any;
export const CalloutContext: any;
export const useCallout: any;
export const stripesShape: any;
export const withStripes: any;
export const useStripes: any;
...
```

All exports from the original modules' `index.js` files are included here.

`stripes-final-form` is the exception to this; since it only has one export, we have created actual type definitions for the [stripesFinalForm function](https://github.com/folio-org/stripes-types/blob/v1.0.1/final-form/lib/stripesFinalForm.d.ts).

### What effect will this PR have?

First off, **this will only affect TypeScript modules**.  Modules which do not use TS will not use the TS compiler and thus these type declaration files are irrelevant.

There are (as far as I know) only two modules which use TypeScript: `ui-calendar` and `ui-bulk-edit`, both of which contain their own type declarations for these modules ([calendar](https://github.com/folio-org/ui-calendar/blob/master/src/typings.d.ts), [bulk-edit](https://github.com/folio-org/ui-bulk-edit/tree/master/src/typings/stripes)) _which take precedence over any we specify here_.  Therefore, we can safely develop these types without breaking any builds; these will only be relied upon downstream once the overridden definitions are removed (which will happen after this PR is merged and additional type development).

@zburke @JohnC-80 if you would please review this when you're able!

### CODE REVIEW

File: components/index.d.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export * from '@folio/stripes-types/components';
```
Comment (zburke) [inline components/index.d.ts:1]: GitHub whines, "Unable to resolve path to module '@folio/stripes-types/components'". Do you understand why? Are you able to import from `@folio/stripes-types/components` locally?

**Review by zburke — COMMENTED** [review]
We need to sort out why GitHub can't resolve `@folio/stripes-types`, and decide where `@types/react` belongs in our dependency hierarchy for this can be merged.

File: package.json [hunk 2]
```diff
@@ -34,6 +35,7 @@
     "react-redux": "~8.0.5"
   },
   "peerDependencies": {
+    "@types/react": "^17.0.2",
```
[the remark this answers is not in the document]
Comment (ncovercash) [reply package.json:38]: I may have just added this in for some local yarn dependency weirdness; will investigate if this is critical

File: components/index.d.ts (hunk 1 above)
Comment (ncovercash) [reply components/index.d.ts:1]: @zburke where do you see this? It builds fine for me and it seems CI is passing (excluding sonar, which doesn't like my fork)

File: .eslintrc [hunk 3]
```diff
@@ -1,3 +1,12 @@
 {
-  "extends": "@folio/eslint-config-stripes"
+  "extends": "@folio/eslint-config-stripes",
+  "overrides": [
+    {
+      "files": ["**/*.d.ts"],
+      "rules": {
+        // eslint does not like .d.ts files without a corresponding .js file
+        "import/no-unresolved": "off"
+      }
```
[the remark this answers is not in the document]
Comment (ncovercash) [reply .eslintrc:9]: Normally the types would live alongside components, and so we would want to indicate errors when the types exist without the source code. This override would only be needed temporarily till `stripes-types` gets completed

**Comment by ncovercash** [comment]
@zburke would we be able to get this merged soon?  I'd like to start using these typings in `ui-calendar`; I've improved a good few of the types, but copying them back into `ui-calendar` would break some things with `ui-bulk-edit` (since both would be declaring different types for `stripes-components`, as happened a few weeks ago)

**Comment by ncovercash** [comment]
@JohnC-80 / @zburke thanks for the reviews; I think there's only one thing left before this can get merged then:

React typings are provided by a package `@types/react`; modules using these types would need to have the same version of `@types/react`, otherwise very weird things occur (for example, if my `ui-module` with `@types/react@17.0.1` uses component `<Button>` from here with `@types/react@17.0.2`, it will fail to compile; the UI module expects the `<Button>` to be a `React.ComponentType`, which it is, but it's the _wrong_ ComponentType since it doesn't strictly equal).  I'm not an expert in yarn's dependency resolution by any means, and especially not when it comes to peer dependencies — what would be the best way to accomplish this?

**Comment by zburke** [comment]
Merged #211 instead

**Pull request closed without merging**